### PR TITLE
[Feature] Optimize DeepSeek's DeepEP on Ascend NPU

### PR DESCRIPTION
**Pull request by iforgetmyname**
## Motivation

Following our roadmap #8004, this is the first pr to support deepep expert parallelism and speed up with some fusion kernels. Now it is possible to activate `--enable-deepep-moe` with `--deepep-mode low_latency` only under pd disaggregation scenario on decode nodes. However, we only support W8A8 int8 quant method currently.

More info about our DeepEP-compatible kernels, check out our roadmap in [sgl-kernel-npu](https://github.com/sgl-project/sgl-kernel-npu/issues/6)

## Modifications

- a new DeepEPMoE impl w/ gemm and swiglu fusion kernels
- gating_topk kernel using fp32 precision to improve accuracy
- introducing fused rms_norm kernel and add_rms_norm kernel
- introducing fused rope kernel under certain condition, waiting to be generalized

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

## Accuracy & Performance

Accuracy with `python3 -m sglang.test.few_shot_gsm8k --num-questions 1318`

image

Performance on 2xA3 PD disaggregation w/o DP-ATTN

image

## Code Format

image

### CODE REVIEW

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 1]
```diff
@@ -1315,8 +1349,189 @@ def forward_deepgemm_masked(
...
+            else:
+                assert shard_id == "w3"
+                expert_data = expert_data.narrow(shard_dim, shard_size, shard_size)
+            expert_data.copy_(loaded_weight)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        topk_weights: torch.Tensor,
+        topk_idx: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (iforgetmyname) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: resolved

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 2]
```diff
@@ -1315,8 +1349,189 @@ def forward_deepgemm_masked(
...
+    def _load_w8a8_scale(
+        self,
+        param: torch.nn.Parameter,
+        loaded_weight: torch.Tensor,
+        weight_name: str,
+        shard_id: str,
+        expert_id: int,
+    ) -> None:
+        param_data = param.data
+        expert_data = param.data[expert_id]
```
[the remark this answers is not in the document]
Comment (iforgetmyname) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: solved

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 3]
```diff
@@ -1315,8 +1349,189 @@ def forward_deepgemm_masked(
...
+        assert self.activation == "silu"
+        output_dtype = torch.bfloat16
+
+        pertoken_scale = hidden_states[1]
+        hidden_states = hidden_states[0]
+
+        group_list_type = 1
+        seg_indptr = seg_indptr.to(torch.int64)
+
+        import torch_npu
```
[the remark this answers is not in the document]
Comment (iforgetmyname) [reply python/sglang/srt/layers/moe/ep_moe/layer.py:749]: this whole class is guarded by _is_npu

File: python/sglang/srt/distributed/parallel_state.py [hunk 4]
```diff
@@ -591,7 +591,7 @@ def _all_gather_into_tensor(self, output: torch.Tensor, input: torch.Tensor):
             )
 
     def all_gather_into_tensor(self, output: torch.Tensor, input: torch.Tensor):
-        if not supports_custom_op():
+        if is_npu() or not supports_custom_op():
```
Comment (Alcanderian) [inline python/sglang/srt/distributed/parallel_state.py]: _is_npu

File: python/sglang/srt/layers/moe/token_dispatcher/deepep.py [hunk 5]
```diff
@@ -150,19 +175,20 @@ def get_deepep_buffer(
...
-            (deepep_mode != DeepEPMode.LOW_LATENCY)
-            and not global_server_args_dict["enable_two_batch_overlap"]
-            and (DeepEPConfig.get_instance().num_sms < total_num_sms // 2)
-        ):
-            logger.warning(
-                f"Only use {DeepEPConfig.get_instance().num_sms} SMs for DeepEP communication. "
-                f"This may result in highly suboptimal performance. "
-                f"Consider using --deepep-config to change the behavior."
-            )
+        if not is_npu():
```
Comment (Alcanderian) [inline python/sglang/srt/layers/moe/token_dispatcher/deepep.py]: _is_npu

File: python/sglang/srt/distributed/parallel_state.py (hunk 4 above)
Comment (iforgetmyname) [reply python/sglang/srt/distributed/parallel_state.py]: done

File: python/sglang/srt/layers/moe/token_dispatcher/deepep.py (hunk 5 above)
Comment (iforgetmyname) [reply python/sglang/srt/layers/moe/token_dispatcher/deepep.py]: done

**Review by Alcanderian — APPROVED** [review]

**Comment by Alcanderian** [comment]
I cannot see any npu deepep calls in this PR. Are they hidden behind deepep-compatible interfaces?

And VLLM test failed

**Comment by iforgetmyname** [comment]
> I cannot see any npu deepep calls in this PR. Are they hidden behind deepep-compatible interfaces?

yes, refer to the [interface](https://github.com/sgl-project/sgl-kernel-npu/blob/20250802/python/deep_ep/deep_ep/buffer.py#L54)

> And VLLM test failed

looks like fixed in #8859 and #8881

**Comment by ping1jing2** [comment]
> I cannot see any npu deepep calls in this PR. Are they hidden behind deepep-compatible interfaces?
> 
> And VLLM test failed

seems like #8837 is the cause. [8837/checks](https://github.com/sgl-project/sglang/pull/8837/checks)

**Comment by ping1jing2** [comment]
> I cannot see any npu deepep calls in this PR. Are they hidden behind deepep-compatible interfaces?
> 
> And VLLM test failed

see #8886
